File: packages/aitrack-lib/src/readers/cursor/cache.ts
import { mkdirSync, readFileSync, renameSync, rmSync, writeFileSync } from 'node:fs';
import { join } from 'node:path';

import { isFiniteNumber, isRecord } from '../../data/guards.js';
import { environmentValue } from '../../env.js';
import { CACHE_DIR } from '../../paths.js';
import { packageVersion } from '../../version.js';

/**
 * Cursor usage is fetched live on every command that shows it. Caching the raw
 * CSV export lets `show` / `usage` / `top` reuse a recent pull
 * instead of making a network round-trip (and trying the whole credential-shape
 * sequence) every single run.
 *
 * The raw CSV is stored rather than the aggregated DayMap so `csv.ts` stays the
 * one parser and a pricing-table change re-aggregates from cache without the
 * network.
 */
const CACHE_FORMAT = 1;
const CACHE_FILE = 'cursor.json';

/** Default max age for a served cache entry, in seconds (6 hours). */
export const DEFAULT_CURSOR_CACHE_TTL_SECONDS = 21_600;

export interface CursorCacheEntry {
  fetchedAt: string;
  csv: string;
  /** Label of the credential shape that produced this CSV, for the next fetch. */
  workingAuthShape?: string;
}

/** TTL for a served Cursor cache entry: `AITRACK_CURSOR_CACHE_TTL` or the default. */
export function cursorCacheTtlSeconds(): number {
  const raw = environmentValue('AITRACK_CURSOR_CACHE_TTL');
  if (raw === undefined) return DEFAULT_CURSOR_CACHE_TTL_SECONDS;
  const parsed = Number(raw);
  return Number.isFinite(parsed) && parsed >= 0 ? parsed : DEFAULT_CURSOR_CACHE_TTL_SECONDS;
}

function cachePath(): string {
  return join(CACHE_DIR, CACHE_FILE);
}

/**
 * The cached CSV export, or null when there is none, it is unreadable, it was
 * written by another format/version, or caching is disabled.
 */
export function readCursorCache(): CursorCacheEntry | null {
  if (environmentValue('AITRACK_NO_CACHE')) return null;

  let parsed: unknown;
  try {
    parsed = JSON.parse(readFileSync(cachePath(), 'utf8'));
  } catch {
    return null;
  }

  if (
    !isRecord(parsed) ||
    parsed.format !== CACHE_FORMAT ||
    // Costs are baked in when the CSV is aggregated, so a pricing/app change
    // must not serve stale dollars.
    parsed.appVersion !== packageVersion() ||
    typeof parsed.csv !== 'string' ||
    typeof parsed.fetchedAt !== 'string'
  ) {
    return null;
  }

  return {
    fetchedAt: parsed.fetchedAt,
    csv: parsed.csv,
    workingAuthShape:
      typeof parsed.workingAuthShape === 'string' ? parsed.workingAuthShape : undefined,
  };
}

/** Age of a cache entry in seconds, or Infinity when its timestamp is unparseable. */
export function cursorCacheAgeSeconds(entry: CursorCacheEntry, now = Date.now()): number {
  const fetchedAtMs = Date.parse(entry.fetchedAt);
  if (!isFiniteNumber(fetchedAtMs)) return Infinity;
  return (now - fetchedAtMs) / 1000;
}

/** Persist a freshly fetched CSV export. A write failure is swallowed — the
 * cache only ever speeds a command up. */
export function writeCursorCache(entry: CursorCacheEntry): void {
  if (environmentValue('AITRACK_NO_CACHE')) return;

  const payload = JSON.stringify({
    format: CACHE_FORMAT,
    appVersion: packageVersion(),
    fetchedAt: entry.fetchedAt,
    csv: entry.csv,
    ...(entry.workingAuthShape !== undefined && { workingAuthShape: entry.workingAuthShape }),
  });
  const target = cachePath();
  const temporaryPath = `${target}.${String(process.pid)}.tmp`;
  try {
    mkdirSync(CACHE_DIR, { recursive: true });
    writeFileSync(temporaryPath, payload, 'utf8');
    renameSync(temporaryPath, target);
  } catch {
    rmSync(temporaryPath, { force: true });
  }
}

File: crates/aitrack-core/src/readers/cursor/cache.rs
//! Cursor CSV export caching.

use crate::env::environment_value;
use crate::paths::cache_dir;
use crate::timezone::machine_timezone;
use crate::version::package_version;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;

const CACHE_FORMAT: u32 = 1;
const CACHE_FILE: &str = "cursor.json";

/// Default TTL for Cursor cache: 6 hours (21600 seconds).
pub const DEFAULT_CURSOR_CACHE_TTL_SECONDS: u64 = 21_600;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CursorCacheEntry {
    #[serde(rename = "fetchedAt")]
    pub fetched_at: String,
    pub csv: String,
    #[serde(rename = "workingAuthShape", skip_serializing_if = "Option::is_none")]
    pub working_auth_shape: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
struct CursorCacheFile {
    format: u32,
    #[serde(rename = "appVersion")]
    app_version: String,
    timezone: String,
    #[serde(rename = "fetchedAt")]
    fetched_at: String,
    csv: String,
    #[serde(rename = "workingAuthShape", skip_serializing_if = "Option::is_none")]
    working_auth_shape: Option<String>,
}

fn cache_path() -> PathBuf {
    cache_dir().join(CACHE_FILE)
}

/// Get the TTL for Cursor cache from AITRACK_CURSOR_CACHE_TTL env var or default.
pub fn cursor_cache_ttl_seconds() -> u64 {
    environment_value("AITRACK_CURSOR_CACHE_TTL")
        .and_then(|s| s.parse::<u64>().ok())
        .unwrap_or(DEFAULT_CURSOR_CACHE_TTL_SECONDS)
}

/// Read the cached Cursor CSV export, or None if absent/invalid.
pub fn read_cursor_cache() -> Option<CursorCacheEntry> {
    if environment_value("AITRACK_NO_CACHE").is_some() {
        return None;
    }

    let contents = fs::read_to_string(cache_path()).ok()?;
    let cache_file: CursorCacheFile = serde_json::from_str(&contents).ok()?;

    // Validate format, version, timezone
    if cache_file.format != CACHE_FORMAT
        || cache_file.app_version != package_version()
        || cache_file.timezone != machine_timezone()
    {
        return None;
    }

    Some(CursorCacheEntry {
        fetched_at: cache_file.fetched_at,
        csv: cache_file.csv,
        working_auth_shape: cache_file.working_auth_shape,
    })
}

/// Age of a cache entry in seconds.
pub fn cursor_cache_age_seconds(entry: &CursorCacheEntry) -> u64 {
    let fetched_at_ms = chrono::DateTime::parse_from_rfc3339(&entry.fetched_at)
        .ok()
        .map(|dt| dt.timestamp_millis())
        .unwrap_or(0);

    let now_ms = chrono::Utc::now().timestamp_millis();
    ((now_ms - fetched_at_ms) / 1000).max(0) as u64
}

/// Persist a freshly fetched CSV export.
pub fn write_cursor_cache(entry: CursorCacheEntry) {
    if environment_value("AITRACK_NO_CACHE").is_some() {
        return;
    }

    let cache_file = CursorCacheFile {
        format: CACHE_FORMAT,
        app_version: package_version().to_string(),
        timezone: machine_timezone(),
        fetched_at: entry.fetched_at,
        csv: entry.csv,
        working_auth_shape: entry.working_auth_shape,
    };

    let payload = match serde_json::to_string(&cache_file) {
        Ok(p) => p,
        Err(_) => return,
    };

    let target = cache_path();
    let temp_path = target.with_extension(format!("{}.tmp", std::process::id()));

    let _ = fs::create_dir_all(cache_dir());
    if fs::write(&temp_path, payload).is_ok() {
        let _ = fs::rename(&temp_path, &target);
    }
    let _ = fs::remove_file(&temp_path);
}
